clippy_lints/src/types.rs

```diff
@@ -238,6 +265,42 @@ fn check_ty(cx: &LateContext<'_, '_>, ast_ty: &hir::Ty, is_local: bool) {
                         );
                         return; // don't recurse into the type
                     }
+                } else if match_def_path(cx.tcx, def_id, &paths::VEC) {
+                    if_chain! {
+                        // Get the _ part of Vec<_>
+                        if let Some(ref last) = last_path_segment(qpath).args;
+                        if let Some(ty) = last.args.iter().find_map(|arg| match arg {
+                            GenericArg::Type(ty) => Some(ty),
+                            GenericArg::Lifetime(_) => None,
+                        });
+                        // ty is now _ at this point
+                        if let TyKind::Path(ref ty_qpath) = ty.node;
+                        let def = cx.tables.qpath_def(ty_qpath, ty.hir_id);
+                        if let Some(def_id) = opt_def_id(def);
+                        if Some(def_id) == cx.tcx.lang_items().owned_box();
+                        // At this point, we know ty is Box<T>, now get T
+                        if let Some(ref last) = last_path_segment(ty_qpath).args;
+                        if let Some(ty) = last.args.iter().find_map(|arg| match arg {
+                            GenericArg::Type(ty) => Some(ty),
+                            GenericArg::Lifetime(_) => None,
+                        });
+                        if let TyKind::Path(ref ty_qpath) = ty.node;
+                        let def = cx.tables.qpath_def(ty_qpath, ty.hir_id);
+                        if let Some(def_id) = opt_def_id(def);
+                        let boxed_type = cx.tcx.type_of(def_id);
+                        if boxed_type.is_sized(cx.tcx.at(ty.span), cx.param_env);
+                        then {
+                            span_lint_and_sugg(
+                                cx,
+                                VEC_BOX_SIZED,
+                                ast_ty.span,
+                                "you seem to be trying to use `Vec<Box<T>>`, but T is Sized. `Vec<T>` is already on the heap, `Vec<Box<T>>` makes an extra allocation.",
+                                "try",
+                                format!("Vec<{}>", boxed_type),
+                                Applicability::MachineApplicable
```

Comment:
This is not a machine applicable suggestion since it _will_ make your code stop compiling.

Use Unspecified for now, perhaps we should add another variant to this enum in rustc.

Reply:
or actually just use MaybeIncorrect

cc @killercup what do you think about a rustfix prompt for "this suggestion will probably make your code stop compiling but you should be able to fix it by following the type errors"

Reply:
I always used `MaybeIncorrect` when there was a high chance, that the suggestion would make the code stop compiling. e.g.
https://github.com/rust-lang/rust-clippy/blob/379c934f3f0c1266e2a4112a4babe7ec0a6368ce/clippy_lints/src/utils/internal_lints.rs#L271-L279

Reply:
I didn't figure out what the MachineApplicable thing *is*, what's the whole Applicability mumbo jumbo?

Reply:
rustfix can automatically apply suggestions

the applicability stuff is about whether or not it _should_. Automatically applying this suggestion isn't enough because this suggestion is incomplete.

Reply:
> @killercup what do you think about a rustfix prompt for "this suggestion will probably make your code stop compiling but you should be able to fix it by following the type errors"

@Manishearth this is something I'd see as a feature of RLS or another interactive tool. Show a non-auto-fixable thing, and get a light bulb button with the option to "dive into fixing this", where you can easily jump around to new compile errors. I'd love to talk about/hack on this some time after Christmas :)

Reply:
Yeah, but for now we could add an `Incomplete` suggestion tag to the compiler and use it! :)